### ISSUE TITLE
[retest] Does the Planar prison quest end happily?

### ISSUE THREAD

**Issue by lynxlynxlynx** [gemrb/gemrb#409]
cygarniczka had trouble in his play and I saw the same bug when investigating. However, with some significant dialog/scripting fixes under our belt, the question is if the bug still exists.

What's the bug? Once you kill everyone and talk to the locked troupe, everyone is supposed to be teleported back to the theatre. If the bug is still present, only the party does and none of the bards.

**Comment by lynxlynxlynx** [comment]
@caillean7 While I'm waiting for weidu, can you check this as well?

**Comment by caillean7** [comment]
I'm pretty sure I played it on my last playthrough with no problems, but I can certainly retry. But the troupe isn't supposed to be teleported back to the theater, just Haer'dalis, if I remember correctly.

**Comment by lynxlynxlynx** [comment]
Well, he's the important one.

**Comment by caillean7** [comment]
Well, at least the other actors go back to Sigil, I think [here](https://www.youtube.com/watch?v=dLr0cfka9f0) you can see that (I wasn't aware that Veriax LPed the classic BG2, awesome). Good, I'm off to testing if H'd gets teleported back alright.

**Comment by caillean7** [comment]
You're right, Haer'Dalis doesn't get teleported back along with the party! I've made a save after killing the warden for testing. I'll take a look into the game scripts.

**Comment by caillean7** [comment]
Responsible for returning everyone back to the theater is CUT17D.bcs, which is started from Raelis' dialogue. It does for H'd the same as for the party members - the only difference is that he is not in the party at that point (not sure if that has any effect - at least, it shouldn't).

```
IF
	True()
THEN
	RESPONSE #100
		CutSceneId(Player1)
		Wait(1)
		ActionOverride("FFActor1",ForceSpell(Myself,DO_NOTHING))  // SPIN905.SPL (No such index)
		ActionOverride("FFActor2",ForceSpell(Myself,DO_NOTHING))  // SPIN905.SPL (No such index)
		ActionOverride("Raelis",ForceSpell(Myself,DO_NOTHING))  // SPIN905.SPL (No such index)
		Wait(3)
		FadeToColor([20.0],0)
		Wait(1)
		LeaveAreaLUAPanic("AR0510","",[884.596],8)
		LeaveAreaLUA("AR0510","",[884.596],8)
		ActionOverride(Player2,LeaveAreaLUA("AR0510","",[796.595],0))
		ActionOverride(Player3,LeaveAreaLUA("AR0510","",[820.628],0))
		ActionOverride(Player4,LeaveAreaLUA("AR0510","",[863.642],0))
		ActionOverride(Player5,LeaveAreaLUA("AR0510","",[919.644],0))
		ActionOverride(Player6,LeaveAreaLUA("AR0510","",[968.632],0))
		ActionOverride("HaerDalis",LeaveAreaLUA("AR0510","",[908.526],0))
		MultiPlayerSync()
		SetMasterArea("AR0500")
		Wait(1)
		FadeFromColor([20.0],0)
		Wait(1)
		ActionOverride("HaerDalis",StartDialogNoSet(Player1))
		EndCutSceneMode()
END

```

**Comment by lynxlynxlynx** [comment]
Thanks & hmm, I wonder why that SetMasterArea call is needed. This could get complicated, but I'll have to check what happens to him during the leave call.

**Comment by lynxlynxlynx** [comment]
> [GameScript/ERROR]: Actionoverride failed for object: 
> [GameScript/DEBUG]: Object: HaerDalis
No wonder ... but let's see why.

**Comment by lynxlynxlynx** [comment]
It's an interesting issue. Player1 runs the cutscene, but he changes areas before H gets to go. So the cutscene is looking for him in the theater, not the prison any more. If he was a global actor already, it wouldn't matter, but he isn't.

It's too central to try to fix properly now, so I think I'll just add an override for this release (having H be the cutscene runner should work).

**Comment by lynxlynxlynx** [comment]
Having Raelis be the runner didn't help, oddly, but just reordering the escapes does.

**Comment by lynxlynxlynx** [comment]
I guess we could implement MC_LIMBO_CREATURE (set on actors switching areas) and check both areas in such cases.

**Comment by lynxlynxlynx** [comment]
Eh, it's trickier to get right. After ensuring the lookup succeeds and he's found in the first area, one realizes that's not enough. The way our movement code is done is that it only sets the target area in the NPC and let's LoadMap do the actual placement later. However, the new map is already loaded from the previous LeaveAreaLUA calls on the PCs, so it shorts and skips PlacePersistents. So he doesn't get actually placed.

**Comment by lynxlynxlynx** [comment]
Note to self, since it may help with fixing this: the originals also marked scriptables that were in cutscenes — they had their own field (CGameSprite::m_inCutScene in Bubb's RE). Currently we store just the runner, but if everyone(?) was tagged, it would be easier to special case the final movement code. A bit overkill for something we're currently avoiding with a simple override script.
